File: P/733_Flood_Fill.java
class Solution {
    int[][] image;
    int newColor;
    int oldColor;
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if(image[sr][sc] == newColor)
            return image;
        
        this.image = image;
        this.newColor = newColor;
        this.oldColor = image[sr][sc];
        
        fill(sr, sc);
        
        return image;
    }
    
    private void fill(int r, int c) {
        if(r < 0 || r >= image.length || c < 0 || c >= image[0].length || image[r][c] != oldColor)
            return;
        
        image[r][c] = newColor;
        
        fill(r - 1, c);
        fill(r + 1, c);
        fill(r, c - 1);
        fill(r, c + 1);
    }
}

File: 주영/Graph/733. Flood Fill.js
var floodFill = function (image, sr, sc, newColor) {
  let oldColor = image[sr][sc];
  DFS(image, sr, sc, oldColor, newColor);
  return image;
};

var DFS = (image, x, y, oldColor, newColor) => {
  if (
    x < 0 ||
    y < 0 ||
    x >= image.length ||
    y >= image[x].length ||
    image[x][y] === newColor ||
    image[x][y] !== oldColor
  )
    return;

  image[x][y] = newColor;
  DFS(image, x + 1, y, oldColor, newColor);
  DFS(image, x, y + 1, oldColor, newColor);
  DFS(image, x - 1, y, oldColor, newColor);
  DFS(image, x, y - 1, oldColor, newColor);
};

let image = [
  [1, 1, 1],
  [1, 1, 0],
  [1, 0, 1],
];
let sr = 1;
let sc = 1;
let newColor = 2;
